### PR TITLE
Manage adlists

### PR DESCRIPTION
**Pull request by DL6ER**
Changes proposed in this pull request:

- Enable/disable all lists from /etc/pihole/adlists.default from the Settings page
- Add new ad lists to user-defined file that will be preserved on updates

screenshot at 2016-12-16 15-35-36

This section is loaded in collapsed form since it needs quite some space of the Settings page...

@pi-hole/dashboard

### CODE REVIEW

**Comment by Mcat12** [comment]
Could we add a text field on the bottom for custom lists? Then they can add them there. Also, how will we handle if the lists are changed in the repo and their `adlists.list` is out of date, but has their changes?

**Comment by DL6ER** [comment]
Sure, I can add a field where a user can add lists. I'd store them in an extra file, so the original `adlists.list` can always be updated, i.e. overwritten, by us when this it is necessary.

**Comment by Mcat12** [comment]
I think it's more of a deeper problem. Since `adlists.list` is copied from `adlists.default`, it isn't updated with the normal update. I like the idea of adding extra lists in a separate file. Then we can change up a bit how we handle list updates. Perhaps only have one `adlists.list` file, which users can modify. On an update, we check the user's list against the updated list and add in the lists that are in the updated version but the the user's version? Then that raises the question of what to do if a list changes from active to commented out or flat deleted in an update. Do we comment out the user's list, delete the list from the file, etc? Or leave it up to the user?

**Comment by DL6ER** [comment]
Currently I do the following: Edit `adlists.default` directly such that it will get stashed and changed how we want it during an update. The `extra.list` would be separate. I'd propose to comment out everything in this file on update. That would lead to the situation that an user always has the configuration we think is the best after an update - however, his custom lists are not deleted, but only disabled for the moment. He can simply go to the Settings page and customize as he likes.

This might be some level of inconvenience connected to each update, but we can be sure that all users have a working situation (that we can control) after each update. Re-enabling the extra lists is done in seconds.

**Comment by DL6ER** [comment]
Updated screenshot in PR description

**Comment by DL6ER** [comment]
@Mcat12 I'd propose to go with `adlists.list` + new `adlists.user`. Not sure what to do with an possibly already existing `adlists.list`, because I'd like to overwrite on every Pi-hole update like:
```
cp /etc/.pihole/adlists.default /etc/pihole/adlists.lists
```
to enforce an up-to-date list.

**Comment by Mcat12** [comment]
Gravity actually already overrides the default list on every run. ;) https://github.com/pi-hole/pi-hole/blob/master/gravity.sh#L335

**Comment by Mcat12** [comment]
That sounds like it should work well. Just will have to try and re-enable/disable the lists that are in the default list based on the user's prior settings, if possible.

**Comment by Mcat12** [comment]
Does this require any Core changes?

File: php/savesettings.php [hunk 1]
```diff
@@ -48,6 +48,36 @@ function validDomain($domain_name)
...
+function readAdlists(&$list, $listname)
+{
+	// Reset list
+	$list = [];
+	$handle = @fopen("/etc/pihole/adlists.".$listname, "r");
+	if ($handle)
+	{
+		while (($line = fgets($handle)) !== false)
+		{
+			if(substr($line, 0, 2) === "#h")
```
Comment (Mcat12) [inline php/savesettings.php]: Could this check for `#http` instead? There might be a comment that starts like `#here's some list...`

File: php/savesettings.php [hunk 2]
```diff
@@ -48,6 +48,36 @@ function validDomain($domain_name)
 			"8.20.247.20" => "Comodo"
 		];
 
+$adlistsdefault = [];
+$adlistsuser = [];
+function readAdlists(&$list, $listname)
+{
+	// Reset list
+	$list = [];
```
Comment (Mcat12) [inline php/savesettings.php]: Why take in a list and just rewrite it? Could this return a list instead?
`$adlistsuser = readAdlists("user");`

**Comment by DL6ER** [comment]
@Mcat12 These are the core changes:
Allow enabling/disabling of adlists via webUI https://github.com/pi-hole/pi-hole/pull/988

File: php/savesettings.php [hunk 3]
```diff
@@ -48,6 +48,37 @@ function validDomain($domain_name)
...
+				array_push($list, [true,rtrim($line)]);
+			}
+		}
+		fclose($handle);
+	}
+	return $list;
+}
+
+	// Read available adlists
+	$adlistsdefault = readAdlists("default");
```
Comment (Mcat12) [inline php/savesettings.php]: Shouldn't this be reading `list` instead of `default`? The web interface is modifying default instead of list.

**Comment by PromoFaux** [comment]
I wonder if this could be expanded to include a description/title for each list? Not sure how that would work off the top of my head, but food for thought...

**Comment by PromoFaux** [comment]
Also, could the save button be expanded to "save and update" so it then loads them into gravity? Saves the user a couple of clicks

**Comment by DL6ER** [comment]
> I wonder if this could be expanded to include a description/title for each list?

Solution:
Add a line staring with `### Some description after three hash marks` **after** the link. It this line is there, I'll take it for the description. If not only the link will be shown.

> Also, could the save button be expanded to "save and update" so it then loads them into gravity?

Okay, this will then redirect the user to the gravity update page and immediately start the update.

**Comment by DL6ER** [comment]
WIP since we might change the format of the adlist to CSV

**Comment by technicalpyro** [comment]
i oticed this thread a few days ago and am wondering if the formatting of my personal adlist could be useful in it 

```
###A word of warning (i.e: Please DO NOT just install all these lists at once):
###Installing all of these lists will result in the Query Log tool of Pi-hole to return an ajax datatables error, when it loads
###If this doesn't deter you, consider skipping a couple of lists known for false positives (e.g: hosts.eladkarako.com)
###If you have hardware with 1GB of RAM or more, you may increase PHP's memory: memory_limit = 180M within /etc/php5/fpm/php.ini OR /etc/php5/cgi/php.ini (FPM if it exists, otherwise CGI, and then sudo service php5-fpm restart OR sudo service php5-cgi restart)
###Raising PHP's memory beyond this suggestion may result in your hardware becoming unresponsive (freezing), as multiple 'children' with this limit are spawned via /etc/php5/fpm/pool.d/www.conf

##Suspicious Lists:

#Bill Stearn
http://v.firebog.net/hosts/BillStearns.txt
#CAMELEON
http://sysctl.org/cameleon/hosts
#DSHEILD Suspicious
https://www.dshield.org/feeds/suspiciousdomains_Low.txt
#Elad Karako
https://raw.githubusercontent.com/eladkarako/hosts.eladkarako.com/master/_raw__hosts.txt
#Joe Wein
http://www.joewein.net/dl/bl/dom-bl-base.txt
#Mahakala
https://adblock.mahakala.is
#hpHosts Spam
https://hosts-file.net/grm.txt
#Andy Short
http://hostsfile.mine.nu/hosts0.txt
#Piwik Referrer Spam
https://raw.githubusercontent.com/piwik/referrer-spam-blacklist/master/spammers.txt
#ReddestDream
https://raw.githubusercontent.com/ReddestDream/reddestdream.github.io/master/Projects/MinimalHosts/etc/MinimalHostsBlocker/minimalhosts
#Steven Blacks KAD
https://raw.githubusercontent.com/StevenBlack/hosts/master/data/KADhosts/hosts
#Steven Blacks Spammers
https://raw.githubusercontent.com/StevenBlack/hosts/master/data/add.Spam/hosts
#SecureMecca
http://securemecca.com/Downloads/hosts.txt
#Someone Who Cares
http://someonewhocares.org/hosts/zero/hosts
#Spam404
https://raw.githubusercontent.com/Dawsey21/Lists/master/main-blacklist.txt
#Vokins yhosts
https://raw.githubusercontent.com/vokins/yhosts/master/hosts
#Winhelp2002 MVPS
http://winhelp2002.mvps.org/hosts.txt

##Advertising Lists:

#AdAway
https://raw.githubusercontent.com/AdAway/adaway.github.io/master/hosts.txt
#Disconnect Ads
https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt
#hpHosts Ads/Trackers
https://hosts-file.net/ad_servers.txt
#Easylist
http://v.firebog.net/hosts/Easylist.txt
#Steven Blacks Spotify Ads
https://raw.githubusercontent.com/StevenBlack/hosts/master/data/SpotifyAds/hosts
#Steven Blacks Unchecky Ads
https://raw.githubusercontent.com/StevenBlack/hosts/master/data/UncheckyAds/hosts

##Tracking & Telemetry Lists:

#Airelle Trackers
http://v.firebog.net/hosts/Airelle-trc.txt
#Easyprivacy
http://v.firebog.net/hosts/Easyprivacy.txt
#Quidsup Trackers
https://raw.githubusercontent.com/quidsup/notrack/master/trackers.txt
#Steven Blacks Trackers
https://raw.githubusercontent.com/StevenBlack/hosts/master/data/add.2o7Net/hosts
#Steven Blacks Microsoft Telemetry
https://raw.githubusercontent.com/StevenBlack/hosts/master/data/tyzbit/hosts

##Malicious Lists:

#Airelle High Risk
http://v.firebog.net/hosts/Airelle-hrsk.txt
#Disconnect Malvertising
https://s3.amazonaws.com/lists.disconnect.me/simple_malvertising.txt
#DNS-BH Malware
http://mirror1.malwaredomains.com/files/justdomains
#hpHosts Exploit
https://hosts-file.net/exp.txt
#hpHosts Malware
https://hosts-file.net/emd.txt
#hpHosts Phishing
https://hosts-file.net/psh.txt
#'Immortal' Malware Domains
https://mirror.cedia.org.ec/malwaredomains/immortal_domains.txt
#Malware Domain List
: https://www.malwaredomainlist.com/hostslist/hosts.txt
#Mandiant APT1 Report
https://bitbucket.org/ethanr/dns-blacklists/raw/8575c9f96e5b4a1308f2f12394abd86d0927a4a0/bad_lists/Mandiant_APT1_Report_Appendix_D.txt
#Openphish
https://openphish.com/feed.txt
#Quidsup Malicious
https://raw.githubusercontent.com/quidsup/notrack/master/malicious-sites.txt
#Ransomware Tracker
https://ransomwaretracker.abuse.ch/downloads/RW_DOMBL.txt
#Steven Blacks Risky Hosts
https://raw.githubusercontent.com/StevenBlack/hosts/master/data/add.Risk/hosts
#Shalla Malicious
http://v.firebog.net/hosts/Shalla-mal.txt
#Zeus Tracker
https://zeustracker.abuse.ch/blocklist.php?download=domainblocklist
```

**Comment by Mcat12** [comment]
image

**Comment by Mcat12** [comment]
image

**Comment by DL6ER** [comment]
Why should the user want to delete them when he can disable them?

**Comment by technicalpyro** [comment]
is there a way to label the lists using a more descriptive title if is is present in the adlists.list file ?

**Comment by PromoFaux** [comment]
@DL6ER some people might just want to have a clean slate, and not use any of the defaults. Why have them all in the list disabled if they're never going to use them?

@technicalpyro possibly, for now I think it's nice just to get the functionality out there and work on the polish separately!

**Review by AzureMarker — APPROVED** [review]

**Review by AzureMarker — APPROVED** [review]

**Comment by eladkarako** [comment]
@technicalpyro 
here are the updated links, <a href="http://hosts.eladkarako.com/">GitHub hosted</a>.

<pre>
HOSTS (HOSTS format)
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts.txt
<sub>also..
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts0.txt
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts_with_localhost.txt
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts0_with_localhost.txt

HOSTS (AdBlock format for browser)
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts_adblock.txt

AdBlock lists
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts_adblock_anti_annoyances_block.txt
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts_adblock_anti_annoyances_block_inline_script.txt
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts_adblock_anti_annoyances_hide.txt
https://raw.githubusercontent.com/eladkarako/hosts/master/build/hosts_adblock_anti_annoyances_style_inject.txt
</sub>
</pre>

**Pull request merged**